ref/src/interpolate.f90

```diff
@@ -0,0 +1,180 @@
+module interpolate
+  implicit none
+
+contains
+
+  subroutine interpolate_allocate(odims, interpFactor, fdims)
+  implicit none
+
+    ! Subroutine for calculating the lower and upper bounds of an
+    ! interpolated array with regard to interpFactor.  The user is
+    ! expected to call this routine, allocate the interpolated array,
+    ! then call subroutine: interpolate_arrray to fill the
+    ! interpolated array.
+    !
+    ! Subroutine parameters:
+    ! originalArray, 2D array to be interpolated
+    ! odims, 4 elements integer array, lower/upper subscripts of original array
+    ! interpFactor, interpolation factor
+    ! odims, 4 elements integer array, lower/upper subscripts of interpolated
+    ! array
+    !
+    ! For example:
+    ! dimension interpArray(:,:)
+    ! call interpolate_allocate(odims,interpFactor,fdims)
+    ! l1=fdims(1)
+    ! u1=fdims(2)
+    ! l2=fdims(3)
+    ! l2=fdims(4)
+    ! allocate (interpArray(l1:u1,l2:u2))
+    ! call interpolate_array(originalArray, odims, interpolatedArray, fdims,
+    ! interpFactor)
+
+    integer, intent(in) :: odims(4)
+    integer, intent(in) :: interpFactor
+    integer, intent(out):: fdims(4)
+
+    fdims(1) = odims(1)
+    fdims(3) = odims(3)
+    fdims(2) = (odims(1)-1) + (odims(2)-odims(1)+1) + (odims(2)-odims(1))*interpFactor
+    fdims(4) = (odims(3)-1) + (odims(4)-odims(3)+1) + (odims(4)-odims(3))*interpFactor
+
+  end subroutine interpolate_allocate
+
+  subroutine interpolate_array(oa, odims, f, fdims, interpFactor)
+
+    ! Bilinear interpolation of f with respect to oa.
+    ! See Wikipedia's description of bilinear interpolation.
+    !    https://en.wikipedia.org/wiki/Bilinear_interpolation
+    !
+    ! This subroutine follows the "Alternative algorithm" in the
```

Comment:
I don't understand the rationale for using the "Alternative algorithm".  It seems more complicated, many more operations, and less straightforward.  Why not use the first one that does the interpolation in one direction, and then computes weights for the second interpolation in the other direction?

Reply:
I thought the "Alternative algorithm" was more straight forward.  The first algorithm uses a matrix vector multiply.  I did not want to put a math library call into the kernel.  And writing out the MxV by hand would have been puzzling to the viewer.  So I went with the "Alternative Algorithm".

Again, no problem switching algorithms.  

Currently, I am still scratching my head about the clumping of the data in the ncview plot of dya.  My interpolation routine produced 3 out for the 4 good plots and passed my unit test.  So I was on the fence about whether it was a bug or not.  I spent Monday writing a Python program to plot the data with matplotlib so I could zoom in to look at the data.  The ncview plot was not able to zoom in to get a closer look.  It looks like a bug in the interpolation, but Hera is closed today.

I moved my ncfiles to Orion.  I will be looking at them to see if I can define the problem more exactly.

Anyway, I could compare the old and new bi-linear routines, see if that reveals the bug.

Reply:
I respectfully disagree.  There is no matrix math involved and no reason to call a math library.  

```
! Get the weights for the x direction
wx1 = (x2 - x) / (x2 - x1)
wx2 = (x - x1) / (x2 - x1)

! Get the weights for the y direction
wy1 = (y2 -y) / (y2 - y1)
wy2 = (y - y1) / (y2 -y1)

! interpolate
f(x, y) = wy1 * (wx1 * Q11 + wx2 * Q21) + wy2 * (wx1 * Q12 + wx2 * Q22)
```

That's it.  Very simple.